Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -158,6 +177,39 @@ theorem bit_eq_zero {n : ℕ} {b : Bool} : n.bit b = 0 ↔ n = 0 ∧ b = false :
   cases b <;> simp [Nat.bit0_eq_zero, Nat.bit1_ne_zero]
 #align nat.bit_eq_zero Nat.bit_eq_zero
 
+theorem bit_ne_zero_iff {n : ℕ} {b : Bool} : n.bit b ≠ 0 ↔ n = 0 → b = true := by
+  simpa only [not_and, Bool.not_eq_false] using (@bit_eq_zero n b).not
+
+/-- An alternative for `bitwise_bit` which replaces the `f false false = false` assumption
+with assumptions that neither `bit a m` nor `bit b n` are `0`
+(albeit, phrased as the implications `m = 0 → a = true` and `n = 0 → b = true`) -/
+lemma bitwise_bit' {f : Bool → Bool → Bool} (a : Bool) (m : Nat) (b : Bool) (n : Nat)
+    (ham : m = 0 → a = true) (hbn : n = 0 → b = true) :
+    bitwise f (bit a m) (bit b n) = bit (f a b) (bitwise f m n) := by
+  conv_lhs => unfold bitwise
+  rw [←bit_ne_zero_iff] at ham hbn
+  simp only [ham, hbn, bit_mod_two_eq_one_iff, Bool.decide_coe, ← div2_val, div2_bit, ne_eq,
+    ite_false]
+  conv_rhs => simp only [bit, bit1, bit0, Bool.cond_eq_ite]
+  split_ifs with hf <;> rfl
+
+lemma bitwise_eq_binaryRec (f : Bool → Bool → Bool) :
+    bitwise f =
+    binaryRec (fun n => cond (f false true) n 0) fun a m Ia =>
+      binaryRec (cond (f true false) (bit a m) 0) fun b n _ => bit (f a b) (Ia n) := by
+  funext x y
+  induction x using binaryRec' generalizing y with
+  | z => simp only [bitwise_zero_left, binaryRec_zero, Bool.cond_eq_ite]
+  | f xb x hxb ih =>
+    rw [←bit_ne_zero_iff] at hxb
+    simp_rw [binaryRec_of_ne_zero _ _ hxb, bodd_bit, div2_bit, eq_rec_constant]
+    induction y using binaryRec' with
+    | z => simp only [bitwise_zero_right, binaryRec_zero, Bool.cond_eq_ite]
+    | f yb y hyb =>
+      rw [←bit_ne_zero_iff] at hyb
+      simp_rw [binaryRec_of_ne_zero _ _ hyb, bitwise_of_ne_zero hxb hyb, bodd_bit, ←div2_val,
+        div2_bit, eq_rec_constant, ih]
```

Comment:
I replaced this proof with one using `binaryRec'`, now that we have `bitwise_bit` proven first.